### PR TITLE
feature: BOA Holder analytics chart added

### PR DESCRIPTION
**Pull request by waqarRns**
update: testcase updated inorder to keep track of BOA holder accurately.

### CODE REVIEW

File: src/Stoa.ts [hunk 1]
```diff
@@ -2019,6 +2021,112 @@ class Stoa extends WebService {
                 return res.status(500).send("Failed to data lookup");
             })
     }
+    /**
+     * GET /average_fee_chart
+     * Called when a request is received through the `/average_fee_chart` handler
+     * The parameter `date` is the start date of the range of dates to look up.
+     * @returns Returns average transaction fee between range (date - filter)
+     */
+    private async boaHolderChart(req: express.Request, res: express.Response) {
```
Comment (TrustHenry) [inline src/Stoa.ts]: The document needs to be updated.

File: src/Stoa.ts [hunk 2]
```diff
@@ -188,6 +189,7 @@ class Stoa extends WebService {
         this.app.post("/preimage_received", this.putPreImage.bind(this));
         this.app.post("/transaction_received", this.putTransaction.bind(this));
         this.app.get("/holders", this.getBoaHolders.bind(this));
+        this.app.get("/holder-chart/:address", this.boaHolderChart.bind(this));
```
Comment (TrustHenry) [inline src/Stoa.ts]:
```suggestion
        this.app.get("/holder_balance_history/:address", this.getHolderBalanceHistory.bind(this));
```

File: src/Stoa.ts (hunk 1 above)
Comment (TrustHenry) [reply src/Stoa.ts]:
```suggestion
     * GET /average_fee_chart
     * Called when a request is received through the `/average_fee_chart` handler
     * The parameter `date` is the start date of the range of dates to look up.
     * @returns Returns average transaction fee between range (date - filter)
     */
    private async getHolderBalanceHistory(req: express.Request, res: express.Response) {   
```

File: src/modules/storage/LedgerStorage.ts [hunk 3]
```diff
@@ -299,8 +299,16 @@ export class LedgerStorage extends Storages {
...
 
+        CREATE TABLE IF NOT EXISTS account_history(
+            address          TEXT,
+            time_stamp       INTEGER NOT NULL,
+            balance          BIGINT(20) UNSIGNED NOT NULL,
+            granularity      TEXT NOT NULL,
+            
+            PRIMARY KEY (address(64), time_stamp, granularity(64))
+        );
+
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts]: https://github.com/bosagora/stoa/blob/v0.x.x/docs/Database_Schema.md
Please add the document.

File: src/modules/storage/LedgerStorage.ts [hunk 4]
```diff
@@ -299,8 +299,16 @@ export class LedgerStorage extends Storages {
...
-        
 
+        CREATE TABLE IF NOT EXISTS account_history(
+            address          TEXT,
+            time_stamp       INTEGER NOT NULL,
+            balance          BIGINT(20) UNSIGNED NOT NULL,
+            granularity      TEXT NOT NULL,
+            
+            PRIMARY KEY (address(64), time_stamp, granularity(64))
+        );
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts]:
```suggestion
        CREATE TABLE IF NOT EXISTS account_history(
            address          TEXT,
            block_height     INTEGER NOT NULL,
            time_stamp       INTEGER NOT NULL,
            balance          BIGINT(20) UNSIGNED NOT NULL,
            granularity      TEXT NOT NULL,
            
            PRIMARY KEY (address(64), block_height)
        );
```

Comment (TrustHenry) [reply src/modules/storage/LedgerStorage.ts]: How about adding an index? (address, time_stamp)

Comment (waqarRns) [reply src/modules/storage/LedgerStorage.ts]: mysql create indexes by default on every column in a composite primary key. so there is no need of creating index again.

File: src/modules/storage/LedgerStorage.ts [hunk 5]
```diff
@@ -299,7 +299,16 @@ export class LedgerStorage extends Storages {
...
+
+        CREATE TABLE IF NOT EXISTS account_history(
+            address          TEXT,
+            block_height     INTEGER     NOT NULL,
+            time_stamp       INTEGER     NOT NULL,
+            balance          BIGINT(20)  UNSIGNED NOT NULL,
+            granularity      TEXT        NOT NULL,
+
+            PRIMARY KEY (address(64), time_stamp, granularity(64))
+        );
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts]:
```suggestion
        CREATE TABLE IF NOT EXISTS account_history(
            address          TEXT,
            time_stamp       INTEGER     NOT NULL,
            granularity      TEXT        NOT NULL,
            block_height     INTEGER     NOT NULL,
            balance          BIGINT(20)  UNSIGNED NOT NULL
            
            PRIMARY KEY (address(64), time_stamp, granularity(64))
        );
```

File: docs/Database_Schema.md [hunk 6]
```diff
@@ -544,3 +544,30 @@ CREATE TABLE IF NOT EXISTS "fees"(
...
+
+### _Schema_
+
+| Column                 | Data Type    | PK | Not NULL | Default  |Description|
+|:-----------------------|:---------    |:--:|:--------:| -------- | --------- |
+|  address               | TEXT         | Y  | Y        |          | Address of Account  |
+|  block_height          | INTEGER      |    | Y        |          | block height  |
+|  time_stamp            | INTEGER      | Y  | Y        |          | unix timestamp  |
+|  balance               | BIGINT(20)   |    | Y        |          | Balance of user at particular time  |
+|  granularity           | BIGINT(20)   | Y  | Y        |          | Total Fee of Transactions in Block  |
```
Comment (TrustHenry) [inline docs/Database_Schema.md]:
```suggestion
|  time_stamp            | INTEGER      | Y  | Y        |          | unix timestamp  |
|  granularity           | BIGINT(20)   | Y  | Y        |          | Total Fee of Transactions in Block  |
|  block_height          | INTEGER      |    | Y        |          | block height  |
|  balance               | BIGINT(20)   |    | Y        |          | Balance of user at particular time  |
```

File: docs/Database_Schema.md [hunk 7]
```diff
@@ -544,3 +544,30 @@ CREATE TABLE IF NOT EXISTS "fees"(
...
+
+### _Create Script_
+```sql
+CREATE TABLE IF NOT EXISTS "account_history"(
+    "address"            TEXT       NOT NULL,
+    "block_height"       INTEGER    NOT NULL,
+    "time_stamp"         INTEGER    NOT NULL,
+    "balance"            BIGINT(20) NOT NULL,
+    "granularity"        BIGINT(20) NOT NULL,
```
Comment (TrustHenry) [inline docs/Database_Schema.md]:
```suggestion
    "address"            TEXT       NOT NULL,
    "time_stamp"         INTEGER    NOT NULL,
    "granularity"        BIGINT(20) NOT NULL,
    "block_height"       INTEGER    NOT NULL,
    "balance"            BIGINT(20) NOT NULL,
```

**Review by TrustHenry — APPROVED** [review]

**Pull request merged**